### PR TITLE
Grease and HRR leak

### LINKED ISSUE #358: ECH contradicts second ClientHello consistency requirements in RFC8446
I just noticed this as I was pondering our various HRR issues. RFC8446, section 4.1.2 says:

> [...] The client will also send a ClientHello when the server has responded to its ClientHello with a HelloRetryRequest.  In that case, the client MUST send the same ClientHello without modification, except as follows:
>
> * [...various cases regarding existing extensions...]
>
> *  Other modifications that may be allowed by an extension defined in the future *and present in the HelloRetryRequest*.

https://tools.ietf.org/html/rfc8446#section-4.1.2

The "and present in the HelloRetryRequest" portion is fun. In RFC8446 as written, we're not allowed to change the `encrypted_client_hello` extension on the second ClientHello unless the server sent `encrypted_client_hello` in HelloRetryRequest, which it doesn't. With the protocol as-is, it seems we'll at least need some text to deal with the contradiction.

The impact is practice is thankfully limited. If the servers does not enforce consistency between the two ClientHellos, it doesn't matter. (They're not required to, but they're not explicitly forbidden from it either, and a stateless server may need to enforce some minimum consistency to avoid getting into a broken state.) If a server does, ECH's behavior on HRR will break it. Such a server is supposed to handshake with ClientHelloOuter, so it just affects the retry flow. That's maybe okay, but mildly annoying. Administrators using server software that checks need to know to first upgrade to a version that doesn't check before trying to deploy ECH.

We can also add an ECH extension to HRR. Then, on ECH-reject + HRR, the server perhaps wouldn't send the extension and the client would be required to replay the extension. This is a waste in many ways, but would work. This would break don't stick out, but some threat models, don't stick out with HRR is hopeless anyway. (See https://github.com/tlswg/draft-ietf-tls-esni/pull/352#issuecomment-721824226.) It would also be weirdly asymmetric between ECH and SH.

That said, this constraint is also pretty obnoxious. For instance, #313 currently adds a new padding extension codepoint, because the old one isn't defined for other messages. However, the RFC8446 allowance only works for the existing padding code point, so you don't want to send any other code points in ClientHello, lest the server force you to keep it unchanged in ClientHello2. So maybe we should confirm no servers did the fully strict check and undo that constraint? (Although, for #313, I personally favor the padding message strategy anyway.)

### PR DESCRIPTION
**Pull request by martinthomson**
But that is just how it is.  We get to choose between that leakage and a
known large amount of bustage.

Closes #358.

### CODE REVIEW

**Review by cbartle891 — APPROVED** [review]

**Comment by cjpatton** [comment]
This solution is that it admits a straightforward, active distinguisher: just intercept CH1, respond with a fake HRR, and look at CH2. It's true that we "give up" on making HRR not stick out in #407. However, one of the design goals there is to prevent active attacks that trigger the HRR code path when it otherwise would not have happened.

**Comment by cbartle891** [comment]
> This solution admits a straightforward, active distinguisher: just intercept CH1, respond with a fake HRR, and look at CH2. It's true that we "give up" on making HRR not stick out in #407. However, one of the design goals there is to prevent active attacks that trigger the HRR code path when it otherwise would not have happened. What we give up on is ensuring "don't stick out" when HRR is inevitable for the given client and server.

Ah, are you saying that this conflicts with #407 (and thus would have to be removed if #407 is merged)?

**Comment by cjpatton** [comment]
I believe #407 resolves the same "OPEN ISSUE", yes.

**Review by chris-wood — APPROVED** [review]

File: draft-ietf-tls-esni.md [hunk 1]
```diff
@@ -802,12 +802,13 @@ structure available for the server, it SHOULD send a GREASE {{?RFC8701}}
...
-
-[[OPEN ISSUE: The above doesn't match HRR handling for either ECH acceptance or
-rejection. See issue https://github.com/tlswg/draft-ietf-tls-esni/issues/358.]]
+When sending a second ClientHello in response to a HelloRetryRequest, the client
+copies the entire "encrypted_client_hello" extension from the first ClientHello.
+The identical value will reveal to an observer that the value of
+"encrypted_client_hello" was fake, but this only occurs if there is a
+HelloRetryRequest.  A client MAY generate a new value, however updating the
+value of the extension is known to cause compatibility issues with some deployed
+implementations.
```
Comment (chris-wood) [inline draft-ietf-tls-esni.md:811]:
```suggestion
HelloRetryRequest. 
```

**Pull request merged**